### PR TITLE
Split interaction menu settings between interaction and self interaction

### PR DESCRIPTION
**Pull request by synixebrett**
**When merged this pull request will:**
 - Title

This is something that's always annoyed me about self interaction, so I figured I'd finally do something about it. Pretty minor but personally not seeing people wave their guns in the air as they navigate self interaction is a decent improvement.

Edit note: By default both backgrounds are set to disabled, my personal setting is the blur screen.

### CODE REVIEW

File: addons/interact_menu/initSettings.sqf [hunk 1]
```diff
@@ -7,3 +7,63 @@
...
 ] call CBA_settings_fnc_init;
+
+[
+    QGVAR(alwaysUseCursorInteraction),
+    "CHECKBOX",
+    localize LSTRING(AlwaysUseCursorInteraction),
+    [format ["ACE %1", localize LSTRING(Category_InteractionMenu)], localize LSTRING(Category_InteractionMenu)],
+    false,
+    false,
+    {GVAR(alwaysUseCursorInteraction) = _this}
```
Comment (Dystopian) [inline addons/interact_menu/initSettings.sqf]: such onChange code is useless, the var is set automatically. Here and below.

Comment (synixebrett) [reply addons/interact_menu/initSettings.sqf]: Cool, I'll take it out.

File: addons/interact_menu/XEH_clientInit.sqf [hunk 2]
```diff
@@ -77,10 +77,24 @@ format ["%1 (%2)", (localize LSTRING(SelfInteractKey)), localize ELSTRING(common
 
 // background options
 ["ace_interactMenuOpened", {
-    if (GVAR(menuBackground)==1) then {[QGVAR(menuBackground), true] call EFUNC(common,blurScreen);};
-    if (GVAR(menuBackground)==2) then {0 cutRsc[QGVAR(menuBackground), "PLAIN", 1, false];};
+    params ["_menuType"];
+    if (_menuType == 0) exitWith {
```
Comment (Dystopian) [inline addons/interact_menu/XEH_clientInit.sqf]: I would remove code duplicating here and below this way (not tested, just concept):
```
private _menuBackgroundSetting = [GVAR(menuBackground), GVAR(menuBackgroundSelf)] select _menuType;
if (_menuBackgroundSetting == 1) exitWith {[QGVAR(menuBackground), true] call EFUNC(common,blurScreen);};
if (_menuBackgroundSetting == 2) exitWith {0 cutRsc [QGVAR(menuBackground), "PLAIN", 1, false];};
```
Note `exitWith` instead of `then`, spaces in syntax and `QGVAR(menuBackground)` for `cutRsc` in both cases which looks like have to be existing class: https://github.com/acemod/ACE3/blob/e5a15d200f44df5fccc0bc5575d18d80b35538dd/addons/interact_menu/CursorMenus.hpp#L9

File: addons/interact_menu/stringtable.xml [hunk 3]
```diff
@@ -416,6 +416,9 @@
             <Chinesesimp>互动选单</Chinesesimp>
             <Chinese>互動選單</Chinese>
         </Key>
+        <Key ID="STR_ACE_Interact_Menu_Category_SelfInteractionMenu">
```
[the remark this answers is not in the document]
Comment (synixebrett) [reply addons/interact_menu/stringtable.xml]: sweet, I didn't even think of looking in there

File: addons/interact_menu/initSettings.sqf [hunk 4]
```diff
@@ -7,3 +7,57 @@
     false,
     {GVAR(selectorColorHex) = _this call BIS_fnc_colorRGBtoHTML} // Stored in Hex to avoid constant conversion
 ] call CBA_settings_fnc_init;
+
+[
+    QGVAR(alwaysUseCursorInteraction),
+    "CHECKBOX",
+    localize LSTRING(AlwaysUseCursorInteraction),
```
Comment (Dystopian) [inline addons/interact_menu/initSettings.sqf]: `LLSTRING` and `LELSTRING` could be used here and below
https://github.com/CBATeam/CBA_A3/blob/6a818a7c9780ce316cfcfad6fbeff199e6c56666/addons/main/script_macros_common.hpp#L1198-L1199

File: addons/interact_menu/initSettings.sqf [hunk 5]
```diff
@@ -7,3 +7,57 @@
...
+    [LSTRING(cursorKeepCentered), LSTRING(cursorKeepCenteredDescription)],
+    [format ["ACE %1", localize LSTRING(Category_InteractionMenu)], localize ELSTRING(Interaction,InteractionMenuSelf)],
+    true,
+    false
+] call CBA_settings_fnc_init;
+
+[
+    QGVAR(menuBackgroundSelf),
+    "LIST",
+    localize LSTRING(background),
```
Comment (dedmen) [inline addons/interact_menu/initSettings.sqf]: afaik CBA settings also calls localize internally if it finds a localize able string.

File: addons/interact_menu/ACE_Settings.hpp [hunk 6]
```diff
@@ -1,26 +1,4 @@
...
-        description = CSTRING(cursorKeepCenteredDescription);
-    };
-    class GVAR(alwaysUseCursorInteraction) {
-        value = 0;
-        typeName = "BOOL";
-        isClientSettable = 1;
-        category = CSTRING(Category_InteractionMenu);
-        displayName = CSTRING(AlwaysUseCursorInteraction);
-    };
     class GVAR(useListMenu) {
```
Comment (dedmen) [inline addons/interact_menu/ACE_Settings.hpp]: might aswell move the other settings to script and get rid of this config. But that might be a different PR

Comment (synixebrett) [reply addons/interact_menu/ACE_Settings.hpp]: I didn't think each setting needed to be separate since most people will probably leave them with the identical defaults.

File: addons/interact_menu/initSettings.sqf (hunk 5 above)
Comment (synixebrett) [reply addons/interact_menu/initSettings.sqf]: Looks like it does, I'll fix that up

File: addons/interact_menu/ACE_Settings.hpp (hunk 6 above)
Comment (Dystopian) [reply addons/interact_menu/ACE_Settings.hpp]: dedmen means just convert other settings to script version, not separating them

Comment (synixebrett) [reply addons/interact_menu/ACE_Settings.hpp]: Ah I see, yeah that can be done

File: addons/interact_menu/initSettings.sqf [hunk 7]
```diff
@@ -1,9 +1,63 @@
...
     [1, 0, 0],
     false,
     {GVAR(selectorColorHex) = _this call BIS_fnc_colorRGBtoHTML} // Stored in Hex to avoid constant conversion
 ] call CBA_settings_fnc_init;
+
+[
+    QGVAR(alwaysUseCursorInteraction),
+    "CHECKBOX",
+    LSTRING(AlwaysUseCursorInteraction),
+    [format ["ACE %1", localize LSTRING(Category_InteractionMenu)], localize LSTRING(Category_InteractionMenu)],
```
Comment (Dystopian) [inline addons/interact_menu/initSettings.sqf]: LLSTRING and LELSTRING here and below

Comment (synixebrett) [reply addons/interact_menu/initSettings.sqf]: Ah right, missed those

**Comment by synixebrett** [comment]
Is there anything outstanding I need to do on this PR? It should it is good for a full review.

File: addons/interact_menu/functions/fnc_renderIcon.sqf [hunk 8]
```diff
@@ -38,7 +38,10 @@ if (_iconFile isEqualTo "") then {
     _iconFile = DEFAULT_ICON;
 };
 
-_text = if (GVAR(UseListMenu)) then {
+_text = if (
+    (GVAR(keyDownSelfAction) && {GVAR(UseListMenuSelf)}) ||
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderIcon.sqf]: lower case `GVAR(useListMenuSelf)` and `GVAR(useListMenu)`

File: addons/interact_menu/functions/fnc_renderIcon.sqf [hunk 9]
```diff
@@ -47,13 +50,21 @@ _text = if (GVAR(UseListMenu)) then {
 [_ctrl, GVAR(iconCount), _text] call FUNC(ctrlSetParsedTextCached);
 GVAR(iconCount) = GVAR(iconCount) + 1;
 
-private _pos = if (GVAR(UseListMenu)) then {
+private _pos = if (
+    (GVAR(keyDownSelfAction) && {GVAR(UseListMenuSelf)}) ||
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderIcon.sqf]: same as above

File: addons/interact_menu/functions/fnc_renderMenu.sqf [hunk 10]
```diff
@@ -90,7 +90,10 @@ if (_numChildren == 1) then {
 private _scaleX = 1;
 private _scaleY = 1;
 
-if (GVAR(UseListMenu)) then {
+if (
+    (GVAR(keyDownSelfAction) && {GVAR(UseListMenuSelf)}) ||
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderMenu.sqf]: same as above

File: addons/interact_menu/functions/fnc_renderSelector.sqf [hunk 11]
```diff
@@ -39,7 +39,11 @@ private _pos = if (GVAR(UseListMenu)) then {
 
 GVAR(iconCount) = GVAR(iconCount) + 1;
 
-if (GVAR(cursorKeepCentered) && {uiNamespace getVariable [QGVAR(cursorMenuOpened),false]}) then {
+if (
+    ((GVAR(keyDownSelfAction) && {GVAR(cursorKeepCenteredSelfInteraction)}) ||
+    {!(GVAR(keyDownSelfAction)) && {GVAR(cursorKeepCentered)}}) &&
+    {uiNamespace getVariable [QGVAR(cursorMenuOpened),false]}
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderSelector.sqf]: space as in `uiNamespace getVariable [QGVAR(cursorMenuOpened), false]`

File: addons/interact_menu/functions/fnc_renderIcon.sqf [hunk 12]
```diff
@@ -38,7 +38,10 @@ if (_iconFile isEqualTo "") then {
     _iconFile = DEFAULT_ICON;
 };
 
-_text = if (GVAR(UseListMenu)) then {
+_text = if (
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderIcon.sqf]: First statement in the same line as opening bracket

File: addons/interact_menu/functions/fnc_renderIcon.sqf [hunk 13]
```diff
@@ -47,13 +50,21 @@ _text = if (GVAR(UseListMenu)) then {
...
+) then {
     [(_sPos select 0) - (0.0095 * SafeZoneW), (_sPos select 1) - (0.0095 * SafeZoneW), 0.20 * SafeZoneW, 0.035 * SafeZoneW]
 } else {
     [(_sPos select 0) - (0.0750 * SafeZoneW), (_sPos select 1) - (0.0095 * SafeZoneW), 0.15 * SafeZoneW, 0.100 * SafeZoneW]
 };
 
-if (GVAR(cursorKeepCentered) && {uiNamespace getVariable [QGVAR(cursorMenuOpened),false]}) then {
+
+if (
+    ((GVAR(keyDownSelfAction) && {GVAR(cursorKeepCenteredSelfInteraction)}) ||
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderIcon.sqf]: First statement in the opening line

File: addons/interact_menu/functions/fnc_renderSelector.sqf [hunk 14]
```diff
@@ -39,7 +39,11 @@ private _pos = if (GVAR(UseListMenu)) then {
 
 GVAR(iconCount) = GVAR(iconCount) + 1;
 
-if (GVAR(cursorKeepCentered) && {uiNamespace getVariable [QGVAR(cursorMenuOpened),false]}) then {
+if (
```
Comment (TheMagnetar) [inline addons/interact_menu/functions/fnc_renderSelector.sqf]: first statement in the opening line

File: addons/interact_menu/functions/fnc_renderIcon.sqf (hunk 12 above)
Comment (Dystopian) [reply addons/interact_menu/functions/fnc_renderIcon.sqf]: then indent would be bad

Comment (synixebrett) [reply addons/interact_menu/functions/fnc_renderIcon.sqf]: leaving it as is, as discussed in slack

File: addons/interact_menu/ACE_Settings.hpp [hunk 15]
```diff
@@ -1,33 +1,4 @@
 class ACE_Settings {
-    class GVAR(alwaysUseCursorSelfInteraction) {
```
Comment (jonpas) [inline addons/interact_menu/ACE_Settings.hpp]: All removed classes need to be re-added with `movedToSQF = 1;` (other entries can be removed).

**Comment by PabstMirror** [comment]
I changed the default of cursorKeepCenteredSelfInteraction to false
this setting is weird if you are not used to it, 
and more importantly it completely fails to work if render scale is not 100%

**Comment by veteran29** [comment]
> and more importantly it completely fails to work if render scale is not 100%

Might be fixed soon:tm:, thanks to @dedmen

**Comment by commy2** [comment]
This merge broke https://github.com/acemod/ACE3/issues/7381
